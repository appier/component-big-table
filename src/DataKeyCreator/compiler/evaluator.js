// evaluator accecpt an AST and variableMap
// and return evaulated value.

const evaluate = node => variable => {

  if(!node){
    throw `Expression is not valid`;
  }

  if(node.type === 'NumberLiteral'){
    return Number(node.value);
  }
  else if(node.type === 'Identifier'){
    //support Immutable
    let ret = null;
    if(typeof variable.get === 'function'){
      ret =  variable.get(node.value);
    }
    else{
      ret = variable[node.value];
    }

    if(ret){
      return ret;
    }
    throw `Unknown identifier: ${node.value}`;
  }
  else if(node.type === 'BinaryExpression'){
    if(node.operator === '+'){
      return evaluate(node.left)(variable) + evaluate(node.right)(variable);
    }
    else if(node.operator === '-'){
      return evaluate(node.left)(variable) - evaluate(node.right)(variable);
    }
    else if(node.operator === '*'){
      return evaluate(node.left)(variable) * evaluate(node.right)(variable);
    }
    else if(node.operator === '/'){
      return evaluate(node.left)(variable) / evaluate(node.right)(variable);
    }
  }
  else {
    throw `Unhandle AST Node: ${node.type}`;
  }
}

const evaluator = ast => variable => {
  return ast.body.map(d => {
    return evaluate(d)(variable);
  });
}

export default evaluator;

