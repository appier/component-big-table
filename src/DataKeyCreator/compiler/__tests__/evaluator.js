import evaluator from '../evaluator';

const data = {
  cost: 100,
  impression: 6000,
  click: 300,
  action: 2,
};

describe('Evaluator', () => {

  it('Number', () => {
    expect(
      evaluator({
        type: 'Program',
        body: [
          {
            type: 'NumberLiteral',
            value: '2'
          }
        ]
      })(data)
    ).toMatchSnapshot();
  });

  it('Identifier', () => {
    expect(
      evaluator({
        type: 'Program',
        body: [
          {
            type: 'Identifier',
            value: 'cost'
          }
        ]
      })(data)
    ).toMatchSnapshot();
  });

  it('BinaryExpression: Number', () => {
    expect(
      evaluator({
        type: 'Program',
        body: [
          {
            type: 'BinaryExpression',
            left: {
              type: 'NumberLiteral',
              value: '3'
            },
            operator: '+',
            right: {
              type: 'NumberLiteral',
              value: '2'
            }
          }
        ]
      })(data)
    ).toMatchSnapshot();
  });

  it('BinaryExpression: Number, Identifier', () => {
    expect(
      evaluator({
        type: 'Program',
        body: [
          {
            type: 'BinaryExpression',
            left: {
              type: 'NumberLiteral',
              value: '1'
            },
            operator: '+',
            right: {
              type: 'Identifier',
              value: 'cost'
            }
          }
        ]
      })(data)
    ).toMatchSnapshot();
  });

  it('BinaryExpression: Identifier, Identifier', () => {
    expect(
      evaluator({
        type: 'Program',
        body: [
          {
            type: 'BinaryExpression',
            left: {
              type: 'Identifier',
              value: 'cost'
            },
            operator: '/',
            right: {
              type: 'Identifier',
              value: 'action'
            }
          }
        ]
      })(data)
    ).toMatchSnapshot();
  });

  it('BinaryExpression: Paren ', () => {
    expect(
      evaluator({
        type: 'Program',
        body: [
          {
            type: 'BinaryExpression',
            operator: '*',
            left: {
              type: 'NumberLiteral',
              value: '1'
            },
            right: {
              type: 'BinaryExpression',
              left: {
                type: 'NumberLiteral',
                value: '2'
              },
              operator: '+',
              right: {
                type: 'NumberLiteral',
                value: "3"
              }
            }
          }
        ]
      })(data)
    ).toMatchSnapshot();
  });

  it('BinaryExpression: Paren 2', () => {
    expect(
      evaluator({
        type: 'Program',
        body: [
          {
            type: 'BinaryExpression',
            operator: '/',
            left: {
              type: 'BinaryExpression',
              left: {
                type: 'Identifier',
                value: 'impression'
              },
              operator: '+',
              right: {
                type: 'Identifier',
                value: 'click'
              }
            },
            right: {
              type: 'BinaryExpression',
              left: {
                type: 'Identifier',
                value: 'cost'
              },
              operator: '+',
              right: {
                type: 'Identifier',
                value: 'action'
              }
            }
          }
        ]
      })(data)
    ).toMatchSnapshot();
  });


});