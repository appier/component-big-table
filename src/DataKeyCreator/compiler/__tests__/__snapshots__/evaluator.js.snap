exports[`Evaluator BinaryExpression: Identifier, Identifier 1`] = `
Array [
  50,
]
`;

exports[`Evaluator BinaryExpression: Number 1`] = `
Array [
  5,
]
`;

exports[`Evaluator BinaryExpression: Number, Identifier 1`] = `
Array [
  101,
]
`;

exports[`Evaluator BinaryExpression: Paren  1`] = `
Array [
  5,
]
`;

exports[`Evaluator BinaryExpression: Paren 2 1`] = `
Array [
  61.76470588235294,
]
`;

exports[`Evaluator Identifier 1`] = `
Array [
  100,
]
`;

exports[`Evaluator Number 1`] = `
Array [
  2,
]
`;
